pkg/apis/k0s.k0sproject.io/v1beta1/api.go

```diff
@@ -95,6 +99,20 @@ func (a *APISpec) getExternalURIForPort(port int) string {
 	return fmt.Sprintf("https://%s:%d", addr, port)
 }
 
+// APIServerAddress returns the address the API is listening on
+func (a *APISpec) APIServerAddress() string {
+	return a.getAPIServerAddress(a.BindAddress)
+}
+
+func (a *APISpec) getAPIServerAddress(address string) string {
+	switch address {
+	case "0.0.0.0":
```

Comment:
this is logically wrong, "0.0.0.0" means we want to listen on all interfaces available, while "localhost" means listening only on loopback iface. What it the logic behind that decision?

Reply:
The reason it is "0.0.0.0" here is because without specifying an interface (as the current version of k0s does), the kube-apiserver listens on 0.0.0.0. So the default is "0.0.0.0" to allow the functionality to remain the same as previous versions.

The reason we return "localhost" in `getAPIServerAddress` follows the same reason as the above, k0s uses a hardcoded "localhost" in the kubeconfig and other certs for the kube-apiserver in the current released version.

With this change, it keeps the default functionality of k0s, while allowing someone to change the bind address and still have the certs and kubeconfig be correct (a kubeconfig pointing at "localhost" when the kube-apiserver is listening on 192.168.1.1 doesn't work).

Reply:
> since you can't connect to 0.0.0.0 directly

well, it turns out that one actually can do that as in many (most?) languages connecting to `0.0.0.0` actually means connecting to localhost. Now that this address is used for kube components which are all Golang, I think it should work if they connect to `0.0.0.0`. 

For reference: https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/net/ipsock_posix.go;l=137-143

With this in mind, everything _should_ work if we use the bind address as the address for other components to connect to the API, right?

Reply:
Huh, I was not aware of that. That is good to know! Since this string is used for health checks and a kubeconfig, an easy test is to take an existing cluster and change the kubeconfig to point to `0.0.0.0`. The certificate will have to be signed with `0.0.0.0` as well, since "localhost" resolves to `127.0.0.1`. Just doing that quick test I outlined, I got the expected certificate error.

I'll do a bit more testing on this and see if this logic can be removed while maintaining how k0s currently functions.

Reply:
@gakio12 hello! do you have any updates on the issue?

Reply:
@soider Apologies for the delay on this, I am expecting to be able to work on this in the next few weeks. I have the changes in place locally, but I have lost my test environment the last two months, so was unable to test (apart from time constraints).

Reply:
My 2 cents here: the actual meaning of "0.0.0.0" depends on how it is used. When used to bind, it means all (or any) interfaces, but when used as a client, it means "this" host (or even more strongly, an "[unspecified](https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/net/ip.go;l=117-121;drc=refs%2Ftags%2Fgo1.17.3)" host, and as pointed above, is interpreted to mean "localhost" or the loopback interface, at least in various cases (including the [apiserver self client](https://github.com/kubernetes/apiserver/blob/7f74d873f30ca894d2caf8eaa3e691a26740a520/pkg/server/config_selfclient.go#L67-L97))
So IMHO, although the address does seem to work (albeit with a failure as it's not in the cert SAN), it seems better to me to keep the old behavior and use "localhost" in this case.